src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -618,42 +618,52 @@ static int CountDigits(int value)
         }
 
         // Some fields that have a reserved spot in the header, may not fit in such field anymore, but they can fit in the
-        // extended attributes. They get collected and saved in that dictionary, with no restrictions.
-        private void CollectExtendedAttributesFromStandardFieldsIfNeeded()
+        // extended attributes. They are always collected or updated in that dictionary, with no restrictions.
+        private void CollectExtendedAttributesFromStandardFields()
         {
-            ExtendedAttributes.Add(PaxEaName, _name);
+            if (!_isPaxEaNameSynced)
+            {
+                ExtendedAttributes[PaxEaName] = Name;
+                _isPaxEaNameSynced = true;
+            }
 
-            if (!ExtendedAttributes.ContainsKey(PaxEaMTime))
+            if (!_isPaxEaMTimeSynced)
             {
-                ExtendedAttributes.Add(PaxEaMTime, TarHelpers.GetTimestampStringFromDateTimeOffset(_mTime));
+                ExtendedAttributes[PaxEaMTime] = TarHelpers.GetTimestampStringFromDateTimeOffset(MTime);
+                _isPaxEaMTimeSynced = true;
             }
 
-            if (!string.IsNullOrEmpty(_gName))
+            if (!_isPaxEaGNameSynced && !string.IsNullOrEmpty(GName))
             {
-                TryAddStringField(ExtendedAttributes, PaxEaGName, _gName, FieldLengths.GName);
+                TryAddStringField(ExtendedAttributes, PaxEaGName, GName, FieldLengths.GName);
```

Comment:
`TryAddStringField` will only change the dictionary element if `Encoding.UTF8.GetByteCount(value) > maxLength` and you are signaling _isPaxEaGNameSynced regardless of that condition.